### DISCUSSION TITLE
Morphometric Calculator (Grid) clarification

### DISCUSSION THREAD

**Discussion by DimitrisTsi** [UMEP-dev/UMEP#314] [Q&A]
Hello
I would like to ask a question regarding the use of non rectangular
 grids.
I want to run the morphometric calculator for the buildings DSM as seen in the attached image for the area denoted by the red polygon.
If i run the morphometric calculator using the polygon in red as the 'vector polygon grid' i get the error
" LINE 216 "if np.sum(dsm_array) == (dsm_array.shape[0] * dsm_array.shape[1] * nd):"
ERROR MESSAGE: unsupported operand type(s) for *: 'int' and 'NoneType' "
So i would like to ask which is the correct way to do this.
Should i mask all values outside the red polygon (assigning them to nan), and run the process using the purple rectangle (it's the red polygons extracted extent) ?
I know this is not a bug but a technical question, however any help would be greatly appreciated!
Thanks!

QGIS version : 3.14
UMEP version : 3.17.3
![Στιγμιότυπο 2021-07-13, 9 49 38 πμ](https://user-images.githubusercontent.com/20373671/125445110-834eb25d-c2ab-47dc-a0c8-c2633ddd9868.png)

**Comment by rarygit** [comment]
As I understand it, Issue #12 (Closed, 18 Nov 2016) has resolved this: 

"The LandCoverFractionGrid is only working with squared (or rectangular) polygon grids allinged with the CRS used. 
This will be fixed so that any shape of grids can be used."
"Resolved. Option to include NoData values included."

https://github.com/UMEP-dev/UMEP/issues/12

From this one could deduce that masking the Land Cover Fraction grid is the decisive dataset in your situation.
If I have time I will try masking the Land Cover Fraction grid  a tutorial dataset.

**Comment by biglimp** [comment]
Sorry for late reply. 

This tool was created long ago and could benefit from some updates. I will see when this can happen.

As it is now, a non-rectangular grid should work and if that is not the case, please submit an issue (or attach your data here) so that we can test. With that said, what this plugin does, is to take the bounding box of your polygon grid and clip your dsm. That means that areas outside is also considered. In a more sophisticated plugin, areas outside should be ignored.

So your question if you can replace with noData. Well, you could but that it is like you have farmlands outside of your polygon which will influence e.g. frontal area index and plan area index.
